dlfcn.c

```diff
@@ -208,6 +208,157 @@ static BOOL MyEnumProcessModules( HANDLE hProcess, HMODULE *lphModule, DWORD cb,
+static char *utf8_encode( const wchar_t *wstr )
+{
+    size_t input_str_len = wcslen( wstr );
+    if ( !wstr || input_str_len == 0 )
+    {
+        return NULL;
+    }
+    int size_needed = WideCharToMultiByte( CP_UTF8, 0, wstr, input_str_len, NULL, 0, NULL, NULL );
+    char *to_return = (char *) malloc( size_needed * sizeof( char ) );
+    if ( !to_return )
+        return NULL;
+    WideCharToMultiByte( CP_UTF8, 0, wstr, input_str_len, to_return, size_needed, NULL, NULL );
+    return to_return;
+}
+
+// Create a converted wide character string from a multibyte character string.
+static wchar_t *utf8_decode( const char *str )
+{
+    size_t input_str_len = strlen( str );
+    if ( !str || input_str_len == 0 )
+    {
+        return NULL;
+    }
+    int size_needed = MultiByteToWideChar( CP_UTF8, 0, str, input_str_len, NULL, 0 );
+    wchar_t *to_return = (wchar_t *) malloc( size_needed * sizeof( wchar_t ) );
+    if ( !to_return )
+        return NULL;
+    MultiByteToWideChar( CP_UTF8, 0, str, input_str_len, to_return, size_needed );
+    return to_return;
+}
+
+void *wdlopen( const wchar_t *file, int mode )
+{
+    HMODULE hModule;
+    UINT uMode;
+
+    error_occurred = FALSE;
+
+    /* Do not let Windows display the critical-error-handler message box */
+    uMode = SetErrorMode( SEM_FAILCRITICALERRORS );
+
+    if( file == 0 )
+    {
+        /* POSIX says that if the value of file is 0, a handle on a global
+         * symbol object must be provided. That object must be able to access
+         * all symbols from the original program file, and any objects loaded
+         * with the RTLD_GLOBAL flag.
+         * The return value from GetModuleHandle( ) allows us to retrieve
+         * symbols only from the original program file. EnumProcessModules() is
+         * used to access symbols from other libraries. For objects loaded
+         * with the RTLD_LOCAL flag, we create our own list later on. They are
+         * excluded from EnumProcessModules() iteration.
+         */
+        hModule = GetModuleHandle( NULL );
+
+        if( !hModule )
+            save_err_str( "(null)" );
+    }
+    else
+    {
+        HANDLE hCurrentProc;
+        DWORD dwProcModsBefore, dwProcModsAfter;
+        wchar_t lpFileName[MAX_PATH];
+        size_t i, len;
+
+        len = wcslen( file );
+
+        if( len >= sizeof( lpFileName ) )
+        {
+            SetLastError( ERROR_FILENAME_EXCED_RANGE );
+            char* fileMultibyte = utf8_encode( file );
+            if ( !fileMultibyte )
+                return NULL;
+            save_err_str( fileMultibyte );
+            free( fileMultibyte );
+            hModule = NULL;
+        }
+        else
+        {
+            /* MSDN says backslashes *must* be used instead of forward slashes. */
+            for( i = 0; i < len; i++ )
+            {
+                if( file[i] == '/' )
+                    lpFileName[i] = '\\';

```

Comment:
And here can be seen reason why copy+paste is a bad thing. You are trying to assign single character to wide character array. To wide character array you should assign wide character value, e.g. `= L'\\';`

Reply:
@pali L`\\` is equivalent to `\\`, since it falls in the ASCII range.

Reply:
They are not equivalent, look e.g. at `sizeof(L'\\')` vs `sizeof('\\')`. In above case it you would get same result, but only because of implicit cast. Those implicit cast cause problems when doing copy+paste.